File: programmers/algorithm_data_structure/dfs_bfs/여행경로.py
from collections import defaultdict 
'''
출처: https://dongdongfather.tistory.com/69
defaultdict는 인자로 넣어준 타입을 기본 value로 초기화해준다.
존재하지 않는 key로 호출해도 기본 value가 반환된다.

dictionay로 graph 개념을 아래와 같이 만들 수 있다.

graph를 bfs에서 사용한다면?
(일단 재귀함수는 안되고 Queue 구조에서)
각 단계에서 footprint의 모든 원소를 제거 후 
검사해서 추가해 넣고
다시 복구하는 방식
'''


def dfs(graph, N, key, footprint):
    if len(footprint) == N + 1:
        return footprint

    for idx, country in enumerate(graph[key]):
        graph[key].pop(idx)

        # 아래처럼 안하고 footprint + [country] 같이 바로 넣으면
        # 아주 미세하게 느린 걸 확인했다.
        tmp = footprint[:]
        tmp.append(country)

        # dfs로 파고 든다.
        ret = dfs(graph, N, country, tmp)

        # 다음 단계에서 잘 쓰도록 다시 복원
        graph[key].insert(idx, country)

        if ret:
            return ret


def best_solution(tickets):
    answer = []

    graph = defaultdict(list)

    # graph 생성
    for ticket in tickets:
        graph[ticket[0]].append(ticket[1])
        graph[ticket[0]].sort()

    return dfs(graph, len(tickets), "ICN", ["ICN"])

# ----------------------------

def get_next(tickets, curr_idxs=None):    
    if curr_idxs is None:
        return [idx for idx in range(len(tickets)) if tickets[idx][0] == 'ICN']
    else:
        return [idx for idx in range(len(tickets))
                if tickets[idx][0] == tickets[curr_idxs[-1]][1]
                and idx not in curr_idxs]


def my_solution(tickets):
    '''
    끝까지 탐색이 가능한 첫 경우를 빠르게 찾아야 하므로
    DFS를 사용한다. 

    'ICN', 'ATL' 같이 매번 상태를 경로로 이어가니 코드가 복잡했다.
    그래서 tickets의 index로 상태를 저장하니 코드가 간단해졌다.
    (이미 방문한 곳은 찾지 않는 걸 구현할 때 따로 리스트를 만들지 않아도 된다.)

    미리 정렬을 해두어 알파벳 순으로 선택하도록 한다.
    스택 구조이므로 내림차순으로 정렬해야 A에 가까울수록 먼저 검사한다.
    '''
    sorted_tickets = sorted(tickets, key=lambda x: x[1], reverse=True)
    stack = [[idx] for idx in get_next(sorted_tickets)]

    while stack:
        curr_idxs = stack.pop()
        if len(curr_idxs) == len(sorted_tickets):
            return ['ICN'] + [sorted_tickets[idx][1] for idx in curr_idxs]
        else:
            next_idxs = get_next(sorted_tickets, curr_idxs)
            stack.extend([curr_idxs+[idx] for idx in next_idxs])

File: programmers/algorithm_data_structure/dfs_bfs/여행경로.java
package algorithm_data_structure.dfs_bfs;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;

public class 여행경로 {
    /**
     * 스택 구조를 이용해 dfs를 구현
     * 다만 매번 방문하지 않은 노드를 찾기 위해
     * 현재 경로의 길이만큼 걸린다는 문제가 있음
     * (better solution 참고)
     */
    public String[] mySolution(String[][] tickets) {
        String[] answer = new String[tickets.length + 1];
        answer[0] = "ICN";
        String[][] sortedTickets  = deepCopyAndSort(tickets, true);
        Stack<List<Integer>> paths = new Stack<>();

        // 시작점 설정
        for (int i = 0; i < sortedTickets.length; i++) {
            if (sortedTickets[i][0].equals("ICN")) {
                List<Integer> first = new ArrayList<>();
                first.add(i);
                paths.add(first);
            }
        }

        // 탐색
        while (paths.peek().size() < tickets.length) {
            List<Integer> currPath = paths.pop();

            for (int i = 0; i < sortedTickets.length; i++) {
                int lastTicketIdx = currPath.get(currPath.size() - 1);
                String recentLocation = sortedTickets[lastTicketIdx][1];
                if (!currPath.contains(i) && recentLocation.equals(sortedTickets[i][0])) {
                    List<Integer> copiedPath = deepCopy(currPath);
                    copiedPath.add(i);
                    paths.add(copiedPath);
                }
            }
        }

        // 정답 형식에 맞게 변환
        for (int i = 0; i < paths.peek().size(); i++) {
            answer[i+1] = sortedTickets[paths.peek().get(i)][1];
        }
        
        return answer;
    }

    /*
     * dfs를 재귀함수로 구현한 풀이
     * 위의 풀이에서 스택을 사용할 경우
     * 실패한 탐색을 되돌리는 과정(방문 초기화)을 구별하는 게 복잡함
     * 
     * 재귀함수로 이용하면 구별이 간단함
     */
    public String[] betterSolution(String[][] tickets) {
        String[] answer = new String[tickets.length + 1];
        answer[0] = "ICN";
        boolean[] visited = new boolean[tickets.length];
        String[][] sortedTickets = deepCopyAndSort(tickets, false);
        List<Integer> ret = null;

        for (int i = 0; i < sortedTickets.length; i++) {
            if (sortedTickets[i][0].equals("ICN")) {
                List<Integer> first = new ArrayList<>();
                first.add(i);
                visited[i] = true;
                ret = dfs(sortedTickets, visited, first);
                if (ret != null) {
                    break;
                }
                visited[i] = false;
            }
        }

        for (int i = 0; i < ret.size(); i++) {
            answer[i+1] = sortedTickets[ret.get(i)][1];
        }

        return answer;
    }

    public List<Integer> dfs(String[][] tickets, boolean[] visited, List<Integer> curr) {
        List<Integer> ret = null;
        if (curr.size() == tickets.length) {
            return curr;
        }

        String lastAirport = tickets[curr.get(curr.size() - 1)][1];
        for (int i = 0; i < tickets.length; i++) {
            if (!visited[i] && tickets[i][0].equals(lastAirport)) {
                List<Integer> next = deepCopy(curr);
                next.add(i);
                visited[i] = true;
                ret = dfs(tickets, visited, next);
                if (ret != null) {
                    break;
                }
                visited[i] = false;
            }
        }

        return ret;
    }

    public List<Integer> deepCopy(List<Integer> origin) {
        List<Integer> copy = new ArrayList<>(origin.size());
        for (int value : origin) {
            copy.add(value);
        }
        return copy;
    }

    public String[][] deepCopyAndSort(String[][] origin, boolean reverse){
        String[][] sortedCopy = new String[origin.length][2];

        for (int i = 0; i < origin.length; i++) {
            sortedCopy[i] = origin[i];
        }

        Arrays.sort(sortedCopy, (o1, o2) -> {
            if (reverse) {
                int compareResult = o2[0].compareTo(o1[0]);
                return (compareResult == 0) ? o2[1].compareTo(o1[1]) : compareResult;
            } else {
                int compareResult = o1[0].compareTo(o2[0]);
                return (compareResult == 0) ? o1[1].compareTo(o2[1]) : compareResult;
            }
        });

        return sortedCopy;
    }

}